docs/guides/implementation-guides/how-to-implement-dependency-injection-with-pure-di.md

```diff
@@ -0,0 +1,207 @@
+{
+    desktop.MainWindow = composition.MainWindow;
+}
+```
+
+This will be useful if the main window will require dependencies to be injected.
+
+Advantages over classical DI container libraries:
+- No explicit initialisation of data contexts is required. Data contexts are configured directly in `.axaml` files according to the MVVM approach.
+- The code is simpler, more compact, and requires less maintenance effort.
+- The main window is created in a pure DI paradigm, and it can be easily supplied with all necessary dependencies via DI as regular types.
+
+## Step 4: Modify App.axaml
+
+Next, modify the `App.axaml` file to use view models according to the MVVM approach via a shared resource:
+
+```xml
+<Application xmlns="https://github.com/avaloniaui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+             xmlns:app="clr-namespace:AvaloniaSimpleApp"
+             x:Class="AvaloniaSimpleApp.App"
+             RequestedThemeVariant="Default">
+
+  <!-- "Default" ThemeVariant follows system theme variant.
+  "Dark" or "Light" are other available options. -->
+  <Application.Styles>
+    <FluentTheme />
+  </Application.Styles>
+  
+  <Application.Resources>
+    <app:Composition x:Key="Composition" />
+  </Application.Resources>
+
+</Application>
+```
+
+This markup fragment
+
+```xml
+<Application.Resources>
+    <app:Composition x:Key="Composition" />
+</Application.Resources>
+```
+
+creates a shared resource of type `Composition` and with key _‘Composition’_, which will be further used as a data context in the views.
+
+## Step 5: Create your views according to MVVM approach
+
+You can now use bindings to model views without even editing the views `.cs` code files, for example for the main window in the example this might look like this:
+
+```xml
+<Window xmlns="https://github.com/avaloniaui"
+        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
+        x:Class="AvaloniaSimpleApp.MainWindow"
+        DataContext="{StaticResource Composition}"
+        xmlns:app="clr-namespace:AvaloniaSimpleApp"
+        x:DataType="app:Composition"
+        Title="{Binding MainViewModel.Title}"
+        Content="{Binding MainViewModel.Greetings}"/>
+```
+
+To use bindings in views:
+
+- You can set a shared resource as a data context
+  
+  `DataContext="{StaticResource Composition}"`
+  
+  The resource with key _"Composition"_ was defined in [step 4](#step-4-modify-appaxaml).
+
+- Specify the data type in the context:
+
+  `xmlns:app="clr-namespace:AvaloniaSimpleApp"`
+
+  `x:DataType="app:Composition"`
+
+- Use the bindings as usual:
+
+  `Title="{Binding MainViewModel.Title}"`
+
+Advantages over classical DI container libraries:
+- The code-behind `.cs` files for views are free of any logic.
```

Comment:
There are many ways to do DI without using a static locator or putting any DI logic on code behind that work in basically every DI framework. The original example code for MSDI mentioned one possible way, it was just removed due to the lack of practical examples in the guide.

https://github.com/cerobe/avalonia-docs/blob/9ea95b7540c6eaf9a59ef88776d543ecc707244a/docs/guides/implementation-guides/how-to-implement-dependency-injection.md?plain=1#L11

Reply:
I don't have much experience using Avalonia, but have had experience with WPF. The main way to work with classic DI containers is to use methods like `GetService<T>()`. With all their faults. It's basically a ServiceLocator. At runtime you can't be sure when something goes wrong and a method like AA throws an exception because you forgot to register some implementation.

I'm not sure I know of a way to use them in bindings without any tricks. In the link you suggested, as far as I understand we just provide access to the container via a static property. This approach does not reduce the dependency of the view layer compared to the current MS DI approach.